### DISCUSSION TITLE
Accessibility & performance

### DISCUSSION THREAD

**Discussion by niutech** [pulkitsharma07/OnlyHumans-Proxy#6] [General]
This should be called _OnlyHealthyHumans-Proxy_. How about visually impaired people? It looks like an accessibility nightmare.

How about performance? Witing for a full page to load headlessly in order to snapshot it and send it as a PNG image to the client is a waste of time. Better use progressive JPEG or even better an optimized AVIF with progressive snapshotting (e.g. every second send an updated image while loading).

How does your project compare with existing projects such as [WRP](https://github.com/tenox7/wrp) or [BrowserBox](https://github.com/BrowserBox/BrowserBox)?

Also, it doesn't prevent from scraping by multimodal LLM-s like [Molmo](https://molmo.allenai.org/).

**Comment by pulkitsharma07** [comment]
> Regarding Accessibility

I can't think of way of providing accessibility features without exposing the site to scraping or bots. If you have any suggestions, please let me know. This particular issue has been raised in multiple forums and https://github.com/pulkitsharma07/OnlyHumans-Proxy/issues/7, I am really interested in solving this, but I don't know how.

> Better use progressive JPEG or even better an optimized AVIF with progressive snapshotting

Since the pages are cached in memory the screenshotting process is done once on initial page load and then every X seconds (i.e. in case a request comes after cache expiry of X seconds). Also in the future the screenshots can be taken in a "build" steps and stored on the disk to be served later when the server is running. This can be done because currently I am focusing on just static websites, so contents can be "prerendered" as `png`.

On taking screenshots on the fly and improving response time, I will explore how to get progressive JPEGs / AVIFs working with puppeteer, from my initial research it doesn't support it natively.

> WRP and BrowserBox

I was not aware about these projects, thanks for bringing them to notice.
WRP seems to be made for a specific niche of using vintage browsers, although I think it also uses puppeteer internally.

BrowserBox is similar yes, but it falls into the "streaming" solution I mentioned in the second [FAQ](https://github.com/pulkitsharma07/OnlyHumans-Proxy?tab=readme-ov-file#faq): it is more resource intensive on servers + won't work with CDNs.

**Reply by niutech** [reply]
As for accessibility, why not use an obscure non-ASCII [Unicode text or pseudoalphabet](https://qaz.wtf/u/convert.cgi?text=unicode+text)?

As for better formats than PNG, Puppeteer [natively supports WebP](https://pptr.dev/api/puppeteer.screenshotoptions), which should be much smaller than PNG.
